server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1027,22 +1027,31 @@ public void cleanup(long repositoryStateId, Version repositoryMetaVersion, Actio
         return Collections.emptyList();
     }
 
-    private DeleteResult cleanupStaleIndices(Map<String, BlobContainer> foundIndices, Set<String> survivingIndexIds) {
-        DeleteResult deleteResult = DeleteResult.ZERO;
+    private void cleanupStaleIndices(Map<String, BlobContainer> foundIndices, Set<String> survivingIndexIds,
+                                             Executor executor, GroupedActionListener<DeleteResult> listener) {
+        final GroupedActionListener<DeleteResult> groupedListener = new GroupedActionListener<>(ActionListener.wrap(deleteResults -> {
+            DeleteResult deleteResult = DeleteResult.ZERO;
+            for (DeleteResult result : deleteResults) {
+                deleteResult = deleteResult.add(result);
+            }
+            listener.onResponse(deleteResult);
+        }, listener::onFailure), foundIndices.size() - survivingIndexIds.size());
+
         try {
             for (Map.Entry<String, BlobContainer> indexEntry : foundIndices.entrySet()) {
                 final String indexSnId = indexEntry.getKey();
-                try {
-                    if (survivingIndexIds.contains(indexSnId) == false) {
-                        logger.debug("[{}] Found stale index [{}]. Cleaning it up", metadata.name(), indexSnId);
-                        deleteResult = deleteResult.add(indexEntry.getValue().delete());
+                executor.execute(ActionRunnable.supply(groupedListener, () -> {
```

Comment:
Thanks @piyushdaftary this looks really nice! I thought about this a little and I think we need to use a pattern like `org.elasticsearch.repositories.blobstore.BlobStoreRepository#fileQueueListener` here.

 The current solution will enqueue a potentially very large number of snapshot delete tasks on the `SNAPSHOT` pool's queue, completely blocking it for other operations. This may be a problem since we for example wish to always be able to run restore operations on this pool as well with reasonable latency. If we enqueue a thousand or more index deletes like in your example, then we'll block the whole queue for many minutes.
If we were to just pull these tasks form a queue in a work-stealing manner like we do in the linked method we'd neatly avoid the issue. Would you be interested in working that into this PR?

Reply:
Thanks @original-brownbear for reviewing the code changes. I have updated PR with code changes to make delete staleIndices to pull tasks from a queue in a work-stealing manner. 
Please Let me know if any more changes needs to be done.